File: 2025_02/leetcode2025-02-26.py
class BrowserHistory:

    def __init__(self, homepage: str):
        self.ptr = 1
        self.web = [homepage]

    def visit(self, url: str) -> None:
        while len(self.web)>self.ptr:
            self.web.pop()
        # self.web = self.web[:self.ptr]
        self.web.append(url)
        self.ptr += 1

    def back(self, steps: int) -> str:
        self.ptr = max(1, self.ptr-steps)
        return self.web[self.ptr-1]

    def forward(self, steps: int) -> str:
        self.ptr = min(len(self.web), self.ptr+steps)
        return self.web[self.ptr-1]

if __name__ == "__main__":
    ops = ["BrowserHistory","visit","visit","visit","back","back","forward","visit","forward","back","back"]
    visits = ["leetcode.com","google.com","facebook.com","youtube.com","linkedin.com"]
    fb = [1,1,1,2,2,7]
    v,f = 0,0
    for op in ops:
        if op == "BrowserHistory":
            bh = BrowserHistory(visits[v])
            v += 1
        elif op == "visit":
            bh.visit(visits[v])
            v += 1
        elif op == "back":
            print(bh.back(fb[f]))
            f += 1
        else:
            print(bh.forward(fb[f]))
            f += 1

File: leetcode2025/2025_02/leetcode2025-02-26.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class BrowserHistory {
private:
    int ptr;
    vector<string> web;
public:
    BrowserHistory(string homepage) {
        ptr = 1;
        web.push_back(homepage);
    }
    
    void visit(string url) {
        // while (web.size()>ptr){
        //     web.erase(web.end()-1);
        // }
        web.resize(ptr);
        web.push_back(url);
        ptr++;
    }
    
    string back(int steps) {
        ptr = max(1, ptr-steps);
        return web[ptr-1];
    }
    
    string forward(int steps) {
        ptr = min((int)web.size(), ptr+steps);
        return web[ptr-1];
    }
};

int main(){
    BrowserHistory bh("leetcode.com");
    vector<string> ops = {"visit","visit","visit","back","back","forward","visit","forward","back","back"};
    vector<string> visits = {"google.com","facebook.com","youtube.com","linkedin.com"};
    vector<int> fb = {1,1,1,2,2,7};
    int v = 0, f = 0;
    for (auto& s : ops){
        if (s == "visit"){
            bh.visit(visits[v]);
            v++;
        }
        else if (s == "back"){
            cout << bh.back(fb[f]) << endl;
            f++;
        }
        else{
            cout << bh.forward(fb[f]) << endl;
            f++;
        }
    }
}
